### ISSUE TITLE
Building a Python Wheel (on Ubuntu) with lower version of Protobuf

### ISSUE THREAD

**Issue by chesterxgchen** [OpenMined/PSI#144]
## Question
What's the approach to build Python wheel with lower version of Protobuf ? 

## Further Information
While trying to integrate with Openmined PSI with NVFLARE, our CI/CD process reported that 

```
openmined-psi 1.0.0 depends on protobuf==4.21.12
    nvflare-nightly 0+untagged.1.g34f97a7 depends on protobuf==3.20.2
```
So we hope to build a wheel that has lower protobuf version requirement. 

Look at the code, we changed the following files

https://github.com/OpenMined/PSI/blob/master/private_set_intersection/python/requirements.in#L7
https://github.com/OpenMined/PSI/blob/master/private_set_intersection/python/BUILD#L105

and use

```
pip-compile requirements.in
```
to generate the requirements.txt

then use the following command to build the wheel: 
```
bazel build -c opt //private_set_intersection/python:wheel
```

The output showed the following warning: 
```
~/projects/PSI/private_set_intersection/python$ bazel build -c opt //private_set_intersection/python:wheel
INFO: Analyzed target //private_set_intersection/python:wheel (135 packages loaded, 2284 targets configured).
INFO: Found 1 target...
INFO: From Compiling src/google/protobuf/message_lite.cc [for tool]:
In file included from /usr/include/string.h:535,
                 from external/com_github_protocolbuffers_protobuf/src/google/protobuf/stubs/port.h:39,
                 from external/com_github_protocolbuffers_protobuf/src/google/protobuf/stubs/common.h:48,
                 from external/com_github_protocolbuffers_protobuf/src/google/protobuf/message_lite.h:46,
                 from external/com_github_protocolbuffers_protobuf/src/google/protobuf/message_lite.cc:36:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'uint8_t* google::protobuf::io::EpsCopyOutputStream::WriteRaw(const void*, int, uint8_t*)' at external/com_github_protocolbuffers_protobuf/src/google/protobuf/io/coded_stream.h:684:16,
    inlined from 'virtual uint8_t* google::protobuf::internal::ImplicitWeakMessage::_InternalSerialize(uint8_t*, google::protobuf::io::EpsCopyOutputStream*) const' at external/com_github_protocolbuffers_protobuf/src/google/protobuf/implicit_weak_message.h:103:28,
    inlined from 'bool google::protobuf::MessageLite::SerializePartialToZeroCopyStream(google::protobuf::io::ZeroCopyOutputStream*) const' at external/com_github_protocolbuffers_protobuf/src/google/protobuf/message_lite.cc:411:30:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:29:33: warning: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' specified size between 18446744071562067968 and 18446744073709551615 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
   29 |   return __builtin___memcpy_chk (__dest, __src, __len,
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
   30 |                                  __glibc_objsize0 (__dest));
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
Target //private_set_intersection/python:wheel up-to-date:
  bazel-bin/private_set_intersection/python/openmined.psi-1.0.0-INTERPRETER-ABI-manylinux2014_x86_64.whl
INFO: Elapsed time: 27.419s, Critical Path: 11.53s
INFO: 737 processes: 27 internal, 709 linux-sandbox, 1 local.
INFO: Build completed successfully, 737 total actions
```
Although it said its successful, but there is no wheel

```
~/projects/PSI/private_set_intersection/python$ ls -l bazel-bin/private_set_intersection/python/openmined.psi-1.0.0-INTERPRETER-ABI-manylinux2014_x86_64.whl
ls: cannot access 'bazel-bin/private_set_intersection/python/openmined.psi-1.0.0-INTERPRETER-ABI-manylinux2014_x86_64.whl': No such file or directory
```

## Screenshots
 ![image](https://user-images.githubusercontent.com/512707/210694549-dde5b91c-6186-4be0-9802-b558d866ae35.png)

## System Information
 - OS:  Ubuntu x86_64 
 - OS Version: 20.04
 - Language Version:  python 3.8.16
 - Package Manager Version:, bazel 6.0.0

## Additional Context
This is part of the effort trying to integrate openmined PSI with NVFLARE in Federated learning.  Although with protobuf version conlicts. My local tests seems to be fine. But the unit tests failed in CI process.  So we are hoping there is a build with lower protobuf requirements. 

Does the PSI python needs such the latest version of protobuf ?

**Comment by s0l0ist** [comment]
@chesterxgchen, thanks for raising this. The steps you've taken to lower the `protobuf` requirement are correct. The requirement is only needed at runtime and isn't used for building the wheel so lowering it should just work. I haven't tested the wheel explicitly to confirm, however.

Regarding the wheel output, the file is located in the parent directory. You're using `ls` in the `private_set_intersection/python` directory and need to be in the root (two levels up). If you run the`bazel` commands from the root of the repo, this helps prevent confusion on where to find artifacts.

<img width="1185" alt="image" src="https://user-images.githubusercontent.com/3663107/210829670-af97c9a8-d71f-45f1-9bdc-c2703fafc0d3.png">

**Comment by s0l0ist** [comment]
Ah yeah that was a minor [breaking change](https://github.com/OpenMined/PSI/blob/master/CHANGES.md) with the `1.0.0` release for Python. The import paths have changed. Please use something like `import private_set_intersection.python as psi`. The python tests have examples.

**Comment by s0l0ist** [comment]
Hmm, try uninstalling from pip, then `bazel clean` then try building and installing the wheel. I don't have a linux machine to test at the moment.

**Comment by s0l0ist** [comment]
I haven't been able to reproduce your exact error, but I did encounter something strange when using `pyenv` to switch to `3.8.16`. I needed to clear the bazel cache located in the home directory of the active user `rm -rf ~/.cache`. Then I was able to successfully re-run the tests and build & install the wheel.

1. Navigate to the root of the repository:`cd PSI`
2. `bazel clean --expunge`
3. `rm -rf ~/.cache`
4. `bazel build -c opt //private_set_intersection/python:wheel`
5. python private_set_intersection/python/rename.py
6. pip install ...

**Comment by s0l0ist** [comment]
It appears that the bazel rule for pybind11 requires a local python to be installed and available. I do not have more insight on this issue. I recall at some point in time needing to install `sudo apt install python3-dev` on Ubuntu 22.04.

**Comment by s0l0ist** [comment]
I don't know how to fix this issue. It still feels like a system / python issue.

What works for me on Ubuntu 22.04 is the following:
1. Install `pyenv`
2. Install the version of python via `pyenv install 3.8.16`
3. Set it to the global interpreter `pyenv global 3.8.16`
4. Run the bazel cleanup steps to purge everything.
5. Run the bazel commands.

**Comment by s0l0ist** [comment]
I'll also relax the protobuf requirements as well.

**Comment by s0l0ist** [comment]
@chesterxgchen 1.0.2 has been published.

**Comment by s0l0ist** [comment]
Aha! Thanks for digging. We publish the wheels on Ubuntu 22, I need to update the job to build on ubuntu 20.04 and change the `manylinux2014` attribute depending on the OS/Glibc.

**Comment by s0l0ist** [comment]
Should be fixed with #150
